<h2 style = "color:blue"> Projeto Mega - V6.JF </h2>
<img src = "imagens/LogoJoao-2019.png" alt="Nada" width=80 height=80>
    <p> Projeto inicial de estudo de Jogos da Mega Senna.</p>
<h2 style = "color:darkred">MAPA dos Sorteios Vencedores por Estados</h2>

In [1]:
# imports iniciais 
import numpy as np
import pandas as pd
import seaborn as sns
from matplotlib import pyplot as plt
from datetime import datetime
import folium
import json

%matplotlib inline

In [2]:
#Dataset
data_path = "PlanilhaBruta01.xlsx"
df = pd.read_excel(data_path)

In [3]:
df.head(8)

Concurso  Data Sorteio  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  \
0       1.0       35135.0       41.0        5.0        4.0       52.0   
1       2.0       35142.0        9.0       39.0       37.0       49.0   
2       3.0       35149.0       36.0       30.0       10.0       11.0   
3       NaN           NaN        NaN        NaN        NaN        NaN   
4       4.0       35156.0        6.0       59.0       42.0       27.0   
5       5.0       35163.0        1.0       19.0       46.0        6.0   
6       6.0       35170.0       19.0       40.0        7.0       13.0   
7       7.0       35177.0       56.0       38.0       21.0       20.0   

   5ª Dezena  6ª Dezena  Arrecadacao_Total  Ganhadores_Sena  \
0       30.0       33.0                0.0              0.0   
1       43.0       41.0                0.0              1.0   
2       29.0       47.0                0.0              2.0   
3        NaN        NaN                NaN              NaN   
4        1.0        5.0                0.0              0.0   
5       16.0        2.0                0.0              0.0   
6       22.0       47.0                0.0              0.0   
7        3.0        5.0                0.0              0.0   

             ...              UF Rateio_Sena  Ganhadores_Quina  Rateio_Quina  \
0            ...             NaN        0.00              17.0      39158.92   
1            ...              PR  2307162.23              65.0      14424.02   
2            ...              RN   391192.51              62.0      10515.93   
3            ...              SP         NaN               NaN           NaN   
4            ...             NaN        0.00              39.0      15322.24   
5            ...             NaN        0.00              98.0       5318.10   
6            ...             NaN        0.00             109.0       7214.66   
7            ...             NaN        0.00             100.0       8746.05   

   Ganhadores_Quadra  Rateio_Quadra  Acumulado Valor_Acumulado  \
0             2016.0         330.21        SIM      1714650.23   
1             4488.0         208.91        NÃO            0.00   
2             4261.0         153.01        NÃO            0.00   
3                NaN            NaN        NaN             NaN   
4             3311.0         180.48        SIM       717080.75   
5             5399.0          96.53        SIM      1342488.85   
6             7147.0         110.03        SIM      2286166.33   
7             5736.0         152.48        SIM      3335692.28   

   Estimativa_Prêmio  Acumulado_Mega_da_Virada  
0                0.0                       0.0  
1                0.0                       0.0  
2                0.0                       0.0  
3                NaN                       NaN  
4                0.0                       0.0  
5                0.0                       0.0  
6                0.0                       0.0  
7                0.0                       0.0  

[8 rows x 21 columns]

In [4]:
df.tail(4)

Concurso  Data Sorteio  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  \
2435    2168.0       43657.0       37.0       25.0       29.0        4.0   
2436    2169.0       43659.0       45.0        7.0       34.0       59.0   
2437    2170.0       43663.0       36.0       21.0       51.0       10.0   
2438    2171.0       43666.0       19.0       12.0       13.0       36.0   

      5ª Dezena  6ª Dezena  Arrecadacao_Total  Ganhadores_Sena  \
2435        1.0       27.0         28687214.5              1.0   
2436       51.0       54.0         24219139.0              0.0   
2437       24.0       38.0         39685793.0              0.0   
2438       55.0       44.0         47121518.5              1.0   

                ...              UF  Rateio_Sena  Ganhadores_Quina  \
2435            ...              PR   8177021.67              66.0   
2436            ...             NaN         0.00              18.0   
2437            ...             NaN         0.00              79.0   
2438            ...              SP  21978571.91             118.0   

      Rateio_Quina  Ganhadores_Quadra  Rateio_Quadra  Acumulado  \
2435      25059.82             4054.0         582.82        NÃO   
2436      77574.63             2145.0         929.96        SIM   
2437      28962.87             4697.0         695.90        SIM   
2438      23023.47             7127.0         544.56        NÃO   

     Valor_Acumulado  Estimativa_Prêmio  Acumulado_Mega_da_Virada  
2435            0.00          2500000.0               57159465.26  
2436     12759137.05         16000000.0               57526924.07  
2437     16973996.11         22000000.0               58129046.82  
2438            0.00          3000000.0               58843986.23  

[4 rows x 21 columns]

In [5]:
df.to_csv("data/planilhaOriginal-V1.csv")

In [6]:
# Filtro - separação dos dados para estudo contendo sorteios e estados 
df2 = df.iloc[:, [0,2,3,4,5,6,7,9,10,11,12]] #Comando para separar as colunas que necessito inicialmente no estudo
df2.head(10)

Concurso  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  5ª Dezena  6ª Dezena  \
0       1.0       41.0        5.0        4.0       52.0       30.0       33.0   
1       2.0        9.0       39.0       37.0       49.0       43.0       41.0   
2       3.0       36.0       30.0       10.0       11.0       29.0       47.0   
3       NaN        NaN        NaN        NaN        NaN        NaN        NaN   
4       4.0        6.0       59.0       42.0       27.0        1.0        5.0   
5       5.0        1.0       19.0       46.0        6.0       16.0        2.0   
6       6.0       19.0       40.0        7.0       13.0       22.0       47.0   
7       7.0       56.0       38.0       21.0       20.0        3.0        5.0   
8       8.0       53.0       17.0       38.0        4.0       47.0       37.0   
9       9.0       55.0       43.0       56.0       54.0        8.0       60.0   

   Ganhadores_Sena Cidade   UF  Rateio_Sena  
0              0.0    NaN  NaN         0.00  
1              1.0    NaN   PR   2307162.23  
2              2.0    NaN   RN    391192.51  
3              NaN    NaN   SP          NaN  
4              0.0    NaN  NaN         0.00  
5              0.0    NaN  NaN         0.00  
6              0.0    NaN  NaN         0.00  
7              0.0    NaN  NaN         0.00  
8              0.0    NaN  NaN         0.00  
9              0.0    NaN  NaN         0.00

In [7]:
df2.to_csv("data/plCompleta-Sl-SortPremioEstados.csv")

In [8]:
# Limpeza inicial de dados NaN
# Identificando NaN
df2.isnull().sum()

Concurso            268
1ª Dezena           268
2ª Dezena           268
3ª Dezena           268
4ª Dezena           268
5ª Dezena           268
6ª Dezena           268
Ganhadores_Sena     268
Cidade             1975
UF                 1689
Rateio_Sena         268
dtype: int64

In [9]:
# filtrando os sorteios nos quais houveram ganhadores 
# Usando drop() onde a coluna Ganhadores_Sena é igual a zero
indice = df2.loc[(df2['Ganhadores_Sena'] == 0)] #Aqui separo os que não ganharam 
df3 = df2.drop(indice.index) # aqui crio novo dataframe somente com concursos ganhadores.
df3.head(3)

Concurso  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  5ª Dezena  6ª Dezena  \
1       2.0        9.0       39.0       37.0       49.0       43.0       41.0   
2       3.0       36.0       30.0       10.0       11.0       29.0       47.0   
3       NaN        NaN        NaN        NaN        NaN        NaN        NaN   

   Ganhadores_Sena Cidade  UF  Rateio_Sena  
1              1.0    NaN  PR   2307162.23  
2              2.0    NaN  RN    391192.51  
3              NaN    NaN  SP          NaN

In [10]:
# Verificando os NaN
df3.isnull().sum()

Concurso           268
1ª Dezena          268
2ª Dezena          268
3ª Dezena          268
4ª Dezena          268
5ª Dezena          268
6ª Dezena          268
Ganhadores_Sena    268
Cidade             309
UF                  23
Rateio_Sena        268
dtype: int64

In [11]:
df3.shape

(773, 11)

<h2 style = "color:darkgreen"> Estratégia de limpeza NaN</h2>
<p> Neste ponto podemos ver que existem valores NaN diferentes para as colunas Cidade e Estados. Neste caso, primeiro fiz a substituição dos valores NaN para as cidade e o estado e então na seguencia fiz a limpeza dos demais valores NaN dos sorteios. <br> 
Os valores NaN foram substituidos pelas letras <strong>"xx"</stong> que significa para este trabalho sem informação.
</p>

In [12]:
# Troca de valores NaN para cidades e estados por xx que significa sem informação
df3.loc[df3['Cidade'].isnull(),'Cidade'] = 'xx'
df3.loc[df3['UF'].isnull(),'UF'] = 'xx'

In [13]:
#verificando o estado dos valores NaN 
df3.isnull().sum()

Concurso           268
1ª Dezena          268
2ª Dezena          268
3ª Dezena          268
4ª Dezena          268
5ª Dezena          268
6ª Dezena          268
Ganhadores_Sena    268
Cidade               0
UF                   0
Rateio_Sena        268
dtype: int64

In [14]:
df3.shape

(773, 11)

In [15]:
# Novo dataframe contendo apenas os valores que nos interessam
df4 = df3.dropna()
df4.head(3)

Concurso  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  5ª Dezena  \
1        2.0        9.0       39.0       37.0       49.0       43.0   
2        3.0       36.0       30.0       10.0       11.0       29.0   
11      11.0       25.0       15.0       58.0       37.0       59.0   

    6ª Dezena  Ganhadores_Sena Cidade  UF  Rateio_Sena  
1        41.0              1.0     xx  PR   2307162.23  
2        47.0              2.0     xx  RN    391192.51  
11       38.0              1.0     xx  RJ  15591365.07

In [16]:
df4.shape

(505, 11)

In [17]:
#criando novo arquivo csv para usos futuros com a planilha limpa sem NaN LP
df4.to_csv('data/ganhadoresPorEstados-LP.csv')

<h1 style = "color:darkblue">Mapa de distribuição de ganhadores:</h1>
<p>Com a planilha de sorteios que geraram ganhadores, serão colocados no mapa do Brasil os estados que geraram jogos vitoriosos durante os 2438 (até a data de extração desta base de dados).
</p>

In [18]:
# criando o mapa do Brasil com o folium e o json estados
br_estados = 'data/Brasil_Estados.geojson'
geo_json_data = json.load(open(br_estados))
mapa = folium.Map(width=800, height=600, location=[-15.77972, -47.92972], zoom_start=4)
folium.GeoJson(geo_json_data).add_to(mapa)

folium.GeoJson(geo_json_data, style_function=lambda feature:{'fillcolor':'green',
                                                             'color':'darkred',
                                                             'weigth': 0.2,}).add_to(mapa)
mapa

In [19]:
# Calculo da frequencia de estados que obtiveram jogos ganhadores da Megasenna
valores1 = pd.value_counts(df4["UF"])
valores1

SP    100
RJ     57
MG     56
PR     54
RS     26
BA     25
DF     25
xx     23
ES     18
GO     15
SC     13
CE     13
PE     12
MT     11
MS     10
PA      9
AM      6
SE      5
MA      5
AL      5
RN      4
PB      4
AC      2
RO      2
PI      2
TO      1
XX      1
RR      1
Name: UF, dtype: int64

<p> Na tabela acima temos as frequencias de vitorias da Megasena em todos os jogos por estados do país. Agora pego um dicionário contendo as coordenadas geograficas de cada estado do pais e transformo em um dataframe para em seguida adicionar os dados ao mapa do Brasil. </p>

In [20]:
#criando o dataframe de coordenadas e frequencia de jogos ganhos por estado

obj = {
    'AC': [ -8.77, -70.55]
  , 'AL': [ -9.71, -35.73]
  , 'AM': [ -3.07, -61.66]
  , 'AP': [  1.41, -51.77]
  , 'BA': [-12.96, -38.51]
  , 'CE': [ -3.71, -38.54]
  , 'DF': [-15.83, -47.86]
  , 'ES': [-19.19, -40.34]
  , 'GO': [-16.64, -49.31]
  , 'MA': [ -2.55, -44.30]
  , 'MT': [-12.64, -55.42]
  , 'MS': [-20.51, -54.54]
  , 'MG': [-18.10, -44.38]
  , 'PA': [ -5.53, -52.29]
  , 'PB': [ -7.06, -35.55]
  , 'PR': [-24.89, -51.55]
  , 'PE': [ -8.28, -35.07]
  , 'PI': [ -8.28, -43.68]
  , 'RJ': [-22.84, -43.15]
  , 'RN': [ -5.22, -36.52]
  , 'RO': [-11.22, -62.80]
  , 'RS': [-30.01, -51.22]
  , 'RR': [  1.89, -61.22]
  , 'SC': [-27.33, -49.44]
  , 'SE': [-10.90, -37.07]
  , 'SP': [-23.55, -46.64]
  , 'TO': [-10.25, -48.25]
}
dfc = pd.DataFrame(data=obj)
dfc

AC     AL     AM     AP     BA     CE     DF     ES     GO     MA  \
0  -8.77  -9.71  -3.07   1.41 -12.96  -3.71 -15.83 -19.19 -16.64  -2.55   
1 -70.55 -35.73 -61.66 -51.77 -38.51 -38.54 -47.86 -40.34 -49.31 -44.30   

   ...       PI     RJ     RN     RO     RS     RR     SC     SE     SP     TO  
0  ...    -8.28 -22.84  -5.22 -11.22 -30.01   1.89 -27.33 -10.90 -23.55 -10.25  
1  ...   -43.68 -43.15 -36.52 -62.80 -51.22 -61.22 -49.44 -37.07 -46.64 -48.25  

[2 rows x 27 columns]

In [21]:
dfc2 = dfc.T
dfc2

0      1
AC  -8.77 -70.55
AL  -9.71 -35.73
AM  -3.07 -61.66
AP   1.41 -51.77
BA -12.96 -38.51
CE  -3.71 -38.54
DF -15.83 -47.86
ES -19.19 -40.34
GO -16.64 -49.31
MA  -2.55 -44.30
MT -12.64 -55.42
MS -20.51 -54.54
MG -18.10 -44.38
PA  -5.53 -52.29
PB  -7.06 -35.55
PR -24.89 -51.55
PE  -8.28 -35.07
PI  -8.28 -43.68
RJ -22.84 -43.15
RN  -5.22 -36.52
RO -11.22 -62.80
RS -30.01 -51.22
RR   1.89 -61.22
SC -27.33 -49.44
SE -10.90 -37.07
SP -23.55 -46.64
TO -10.25 -48.25

In [22]:
dfc2.rename(columns={0:"La", 1:"Lo"}, inplace=True)

In [23]:
dfc2

La     Lo
AC  -8.77 -70.55
AL  -9.71 -35.73
AM  -3.07 -61.66
AP   1.41 -51.77
BA -12.96 -38.51
CE  -3.71 -38.54
DF -15.83 -47.86
ES -19.19 -40.34
GO -16.64 -49.31
MA  -2.55 -44.30
MT -12.64 -55.42
MS -20.51 -54.54
MG -18.10 -44.38
PA  -5.53 -52.29
PB  -7.06 -35.55
PR -24.89 -51.55
PE  -8.28 -35.07
PI  -8.28 -43.68
RJ -22.84 -43.15
RN  -5.22 -36.52
RO -11.22 -62.80
RS -30.01 -51.22
RR   1.89 -61.22
SC -27.33 -49.44
SE -10.90 -37.07
SP -23.55 -46.64
TO -10.25 -48.25

In [24]:
dfc2.to_csv("data/coordenadasEstadosBrasil.csv")

In [25]:
dfr = pd.DataFrame(data=valores1)
dfr

UF
SP  100
RJ   57
MG   56
PR   54
RS   26
BA   25
DF   25
xx   23
ES   18
GO   15
SC   13
CE   13
PE   12
MT   11
MS   10
PA    9
AM    6
SE    5
MA    5
AL    5
RN    4
PB    4
AC    2
RO    2
PI    2
TO    1
XX    1
RR    1

In [26]:
#Trocando o nome da coluna de UF para Frq
dfr.rename(columns={"UF":"Frq"}, inplace=True)
dfr

Frq
SP  100
RJ   57
MG   56
PR   54
RS   26
BA   25
DF   25
xx   23
ES   18
GO   15
SC   13
CE   13
PE   12
MT   11
MS   10
PA    9
AM    6
SE    5
MA    5
AL    5
RN    4
PB    4
AC    2
RO    2
PI    2
TO    1
XX    1
RR    1

In [27]:
#Criando um data frame com os dados de AP para somar ao data frame de frequencias
data = {"AP":[0]}
dfe = pd.DataFrame(data)
dfe2 = dfe.T
dfe2.rename(columns={0:"Frq"}, inplace=True)
dfe2

Frq
AP    0

In [28]:
#inserido o Estado de AP que não apresenta ganhadores 
dff = pd.concat([dfr,dfe2])
dff

Frq
SP  100
RJ   57
MG   56
PR   54
RS   26
BA   25
DF   25
xx   23
ES   18
GO   15
SC   13
CE   13
PE   12
MT   11
MS   10
PA    9
AM    6
SE    5
MA    5
AL    5
RN    4
PB    4
AC    2
RO    2
PI    2
TO    1
XX    1
RR    1
AP    0

In [29]:
#ordenando os valores dos indices dos estados
dff = dff.sort_index(ascending=True)
dff

Frq
AC    2
AL    5
AM    6
AP    0
BA   25
CE   13
DF   25
ES   18
GO   15
MA    5
MG   56
MS   10
MT   11
PA    9
PB    4
PE   12
PI    2
PR   54
RJ   57
RN    4
RO    2
RR    1
RS   26
SC   13
SE    5
SP  100
TO    1
XX    1
xx   23

In [30]:
print(dfc2.shape)# Dataframe com coordenadas dos estados
print(dff.shape)# Dataframe com a frequencia dos ganhadores

(27, 2)
(29, 1)


In [31]:
#Elimiando os dados XX e xx do dataframe dff 
dff.drop("XX", inplace=True)
dff.drop("xx", inplace=True)
dff.shape

(27, 1)

In [32]:
#Criando o dataframe contendo as coordenadas de cada estado com a frequencia de ganhadores de cada estado.
dfm = dfc2
dfm["Frq"]=dff.values
dfm

La     Lo  Frq
AC  -8.77 -70.55    2
AL  -9.71 -35.73    5
AM  -3.07 -61.66    6
AP   1.41 -51.77    0
BA -12.96 -38.51   25
CE  -3.71 -38.54   13
DF -15.83 -47.86   25
ES -19.19 -40.34   18
GO -16.64 -49.31   15
MA  -2.55 -44.30    5
MT -12.64 -55.42   56
MS -20.51 -54.54   10
MG -18.10 -44.38   11
PA  -5.53 -52.29    9
PB  -7.06 -35.55    4
PR -24.89 -51.55   12
PE  -8.28 -35.07    2
PI  -8.28 -43.68   54
RJ -22.84 -43.15   57
RN  -5.22 -36.52    4
RO -11.22 -62.80    2
RS -30.01 -51.22    1
RR   1.89 -61.22   26
SC -27.33 -49.44   13
SE -10.90 -37.07    5
SP -23.55 -46.64  100
TO -10.25 -48.25    1

In [33]:
#Na concatenação dos dataframes vi que os valores do RS, RR, MG e MT estavam trocados dai fiz a correção abaixo
dfm.set_value("RS", "Frq", 26)
dfm.set_value("MG", "Frq", 56)
dfm.set_value("RR", "Frq", 1)
dfm.set_value("MT", "Frq", 11)

/home/joaoflorentino/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  
/home/joaoflorentino/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  This is separate from the ipykernel package so we can avoid doing imports until
/home/joaoflorentino/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  after removing the cwd from sys.path.
/home/joaoflorentino/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  """


La     Lo  Frq
AC  -8.77 -70.55    2
AL  -9.71 -35.73    5
AM  -3.07 -61.66    6
AP   1.41 -51.77    0
BA -12.96 -38.51   25
CE  -3.71 -38.54   13
DF -15.83 -47.86   25
ES -19.19 -40.34   18
GO -16.64 -49.31   15
MA  -2.55 -44.30    5
MT -12.64 -55.42   11
MS -20.51 -54.54   10
MG -18.10 -44.38   56
PA  -5.53 -52.29    9
PB  -7.06 -35.55    4
PR -24.89 -51.55   12
PE  -8.28 -35.07    2
PI  -8.28 -43.68   54
RJ -22.84 -43.15   57
RN  -5.22 -36.52    4
RO -11.22 -62.80    2
RS -30.01 -51.22   26
RR   1.89 -61.22    1
SC -27.33 -49.44   13
SE -10.90 -37.07    5
SP -23.55 -46.64  100
TO -10.25 -48.25    1

In [34]:
dfm.to_csv("data/coordenadasComFrequeciaVitorias.csv")

<h2 style = "color:darkblue">Contruindo mapa com informações:</h2>
<p>Com o dataframe dfm contendo as informaçẽos de latitude e longitudes e a frequencia de vitorias de cada estado, os mesmos serão adicionados ao mapa com as devidas marcações.
</p>

In [35]:
# criando o mapa do Brasil com o folium e o json estados e colocando as marcações dos ganhadores 
br_estados = 'data/Brasil_Estados.geojson'
geo_json_data = json.load(open(br_estados))
mapa2 = folium.Map(width=800, height=600, location=[-15.77972, -47.92972],tiles="Stamen Terrain", zoom_start=4)
folium.GeoJson(geo_json_data).add_to(mapa2)

folium.GeoJson(geo_json_data, style_function=lambda feature:{'fillcolor':'green',
                                                             'color':'darkred',
                                                             'weigth': 0.2,}).add_to(mapa2)
lat = dfm['La'].values #aqui lemos os valores da latatitude do dataframe dfm
long = dfm['Lo'].values #aqui lemos os valores da longitude do dataframe dfm
freq = dfm['Frq'].values #aqui lemos os valores da Frequencia de ganhadores do dataframe dfm

for la,lo,frq in zip(lat,long,freq):
    folium.Marker([la,lo], popup=frq, tootip="Clique para mais informações").add_to(mapa2)


mapa2